secure-device-storage/sds-android/securedevicestorage/src/main/java/de/adorsys/android/securedevicestorage/KeystoreTool.java

```diff
@@ -0,0 +1,250 @@
+        CipherOutputStream cipherOutputStream = new CipherOutputStream(
+                outputStream, input);
+        cipherOutputStream.write(plainMessage.getBytes("UTF-8"));
+        cipherOutputStream.close();
+
+        byte[] values = outputStream.toByteArray();
+        return Base64.encodeToString(values, Base64.DEFAULT);
+    }
+
+    @Nullable
+    static String decryptMessage(@NonNull Context context, @NonNull String encryptedMessage) throws
+            NoSuchPaddingException, NoSuchAlgorithmException, NoSuchProviderException,
+            IOException, UnrecoverableEntryException, KeyStoreException,
+            CertificateException, InvalidKeyException {
+
+        Cipher output;
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2
+                && Build.VERSION.SDK_INT < M) {
+            output = Cipher.getInstance(KEY_TRANSFORMATION_ALGORITHM, KEY_CIPHER_JELLYBEAN_PROVIDER);
+        } else if (Build.VERSION.SDK_INT >= M) {
+            output = Cipher.getInstance(KEY_TRANSFORMATION_ALGORITHM, KEY_CIPHER_MARSHMALLOW_PROVIDER);
+        } else {
+            return null;
+        }
+
+        output.init(Cipher.DECRYPT_MODE, getPrivateKey(context));
+
+        CipherInputStream cipherInputStream = new CipherInputStream(
+                new ByteArrayInputStream(Base64.decode(encryptedMessage, Base64.DEFAULT)), output);
+        ArrayList<Byte> values = new ArrayList<>();
+        int nextByte;
+        while ((nextByte = cipherInputStream.read()) != -1) {
+            values.add((byte) nextByte);
+        }
+
+        byte[] bytes = new byte[values.size()];
+        for (int i = 0; i < bytes.length; i++) {
+            bytes[i] = values.get(i);
+        }
+
+        return new String(bytes, 0, bytes.length, "UTF-8");
+    }
+
+    @RequiresApi(api = M)
+    private static void generateMarshmallowKeyPair()
+            throws NoSuchProviderException, NoSuchAlgorithmException,
+            InvalidAlgorithmParameterException {
+
+        Calendar start = Calendar.getInstance();
+        Calendar end = Calendar.getInstance();
+        end.add(Calendar.MONTH, 1);
+
+        KeyGenParameterSpec spec = new KeyGenParameterSpec.Builder(
+                KEY_ALIAS, KeyProperties.PURPOSE_SIGN)
+                .setKeyValidityStart(start.getTime())
+                .setKeyValidityEnd(end.getTime())
+                .setCertificateSerialNumber(BigInteger.ONE)
+                .setCertificateSubject(new X500Principal(KEY_X500PRINCIPAL))
+                .setDigests(KeyProperties.DIGEST_SHA256, KeyProperties.DIGEST_SHA512)
+                .setSignaturePaddings(KeyProperties.SIGNATURE_PADDING_RSA_PSS)
+                .build();
+
+        KeyPairGenerator generator
+                = KeyPairGenerator.getInstance(KEY_ENCRYPTION_ALGORITHM, KEY_KEYSTORE_NAME);
+        generator.initialize(spec);
+
+        generator.generateKeyPair();
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR2)
```

Comment:
Fixed in commit: https://github.com/adorsys/secure-key-storage/pull/54/commits/825b5723dd621d327ac110cc7399f8372493cca4